File: servidor-java/src/link/Crc32.java
package link;


public final class Crc32 {
    private static final String GENERATOR_BITS =
        "100000100110000010001110110110111";
    private static final int CRC_SIZE = 32;

    private Crc32() {
    }

    private static void validateBits(
        String bits,
        String fieldName
    ) {
        if (bits == null || bits.isEmpty()) {
            throw new IllegalArgumentException(
                fieldName + " no puede estar vacío."
            );
        }

        for (
            int currentBitPosition = 0;
            currentBitPosition < bits.length();
            currentBitPosition++
        ) {
            char currentBit = bits.charAt(currentBitPosition);

            if (currentBit != '0' && currentBit != '1') {
                throw new IllegalArgumentException(
                    fieldName + " solo puede contener 0 y 1."
                );
            }
        }
    }

    private static String divide(String dividendBits) {
        char[] workingDividendBits =
            dividendBits.toCharArray();
        int generatorSize = GENERATOR_BITS.length();
        int lastPossibleDivisionPosition =
            dividendBits.length() - generatorSize;

        for (
            int currentDivisionPosition = 0;
            currentDivisionPosition
                <= lastPossibleDivisionPosition;
            currentDivisionPosition++
        ) {
            char currentDividendBit =
                workingDividendBits[currentDivisionPosition];

            if (currentDividendBit == '0') {
                continue;
            }

            for (
                int currentGeneratorPosition = 0;
                currentGeneratorPosition < generatorSize;
                currentGeneratorPosition++
            ) {
                int currentDividendPosition =
                    currentDivisionPosition
                    + currentGeneratorPosition;

                char dividendBit = workingDividendBits[
                    currentDividendPosition
                ];
                char generatorBit = GENERATOR_BITS.charAt(
                    currentGeneratorPosition
                );

                if (dividendBit == generatorBit) {
                    workingDividendBits[
                        currentDividendPosition
                    ] = '0';
                } else {
                    workingDividendBits[
                        currentDividendPosition
                    ] = '1';
                }
            }
        }

        int remainderStartPosition =
            workingDividendBits.length - CRC_SIZE;

        return new String(
            workingDividendBits,
            remainderStartPosition,
            CRC_SIZE
        );
    }

    private static String calculateRemainder(
        String dataBits
    ) {
        validateBits(dataBits, "Los datos");

        String appendedZeros = "0".repeat(CRC_SIZE);
        String dividendBits = dataBits + appendedZeros;

        String remainderBits = divide(dividendBits);
        return remainderBits;
    }

    public static String encode(String dataBits) {
        String remainderBits =
            calculateRemainder(dataBits);
        String codewordBits = dataBits + remainderBits;

        return codewordBits;
    }

    public static boolean verify(String codewordBits) {
        validateBits(
            codewordBits,
            "La palabra código"
        );

        if (codewordBits.length() <= CRC_SIZE) {
            throw new IllegalArgumentException(
                "La palabra código debe contener datos "
                + "y 32 bits de CRC."
            );
        }

        String calculatedRemainderBits =
            divide(codewordBits);
        String expectedRemainderBits =
            "0".repeat(CRC_SIZE);

        return calculatedRemainderBits.equals(
            expectedRemainderBits
        );
    }

    public static String recoverData(
        String codewordBits
    ) {
        boolean codewordIsValid = verify(codewordBits);

        if (!codewordIsValid) {
            throw new IllegalArgumentException(
                "CRC-32 detectó un error "
                + "en la palabra código."
            );
        }

        int dataEndPosition =
            codewordBits.length() - CRC_SIZE;
        String originalDataBits = codewordBits.substring(
            0,
            dataEndPosition
        );

        return originalDataBits;
    }
}

File: cliente-python/src/link/crc32.py
class Crc32:
    GENERATOR_BITS = "100000100110000010001110110110111"
    CRC_SIZE = 32

    @staticmethod
    def _validate_bits(bits: str, field_name: str):
        if not bits:
            raise ValueError(
                f"{field_name} no puede estar vacío."
            )

        for bit in bits:
            if bit != "0" and bit != "1":
                raise ValueError(
                    f"{field_name} solo puede contener 0 y 1."
                )

    @staticmethod
    def _divide(dividend_bits: str) -> str:
        working_dividend_bits = list(dividend_bits)
        generator_size = len(Crc32.GENERATOR_BITS)
        last_possible_division_position = (
            len(dividend_bits) - generator_size
        )

        for current_division_position in range(
            last_possible_division_position + 1
        ):
            current_dividend_bit = working_dividend_bits[
                current_division_position
            ]

            if current_dividend_bit == "0":
                continue

            for current_generator_position in range(
                generator_size
            ):
                current_dividend_position = (
                    current_division_position
                    + current_generator_position
                )

                dividend_bit = working_dividend_bits[
                    current_dividend_position
                ]
                generator_bit = Crc32.GENERATOR_BITS[
                    current_generator_position
                ]

                if dividend_bit == generator_bit:
                    working_dividend_bits[
                        current_dividend_position
                    ] = "0"
                else:
                    working_dividend_bits[
                        current_dividend_position
                    ] = "1"

        remainder_bits = working_dividend_bits[
            -Crc32.CRC_SIZE:
        ]
        return "".join(remainder_bits)

    @staticmethod
    def _calculate_remainder(data_bits: str) -> str:
        Crc32._validate_bits(data_bits, "Los datos")

        appended_zeros = "0" * Crc32.CRC_SIZE
        dividend_bits = data_bits + appended_zeros

        remainder_bits = Crc32._divide(dividend_bits)
        return remainder_bits

    @staticmethod
    def encode(data_bits: str) -> str:
        remainder_bits = Crc32._calculate_remainder(data_bits)
        codeword_bits = data_bits + remainder_bits

        return codeword_bits

    @staticmethod
    def verify(codeword_bits: str) -> bool:
        Crc32._validate_bits(
            codeword_bits,
            "La palabra código",
        )

        if len(codeword_bits) <= Crc32.CRC_SIZE:
            raise ValueError(
                "La palabra código debe contener datos "
                "y 32 bits de CRC."
            )

        calculated_remainder_bits = Crc32._divide(
            codeword_bits
        )
        expected_remainder_bits = "0" * Crc32.CRC_SIZE

        return (
            calculated_remainder_bits
            == expected_remainder_bits
        )

    @staticmethod
    def recover_data(codeword_bits: str) -> str:
        codeword_is_valid = Crc32.verify(codeword_bits)

        if not codeword_is_valid:
            raise ValueError(
                "CRC-32 detectó un error en la palabra código."
            )

        data_end_position = (
            len(codeword_bits) - Crc32.CRC_SIZE
        )
        original_data_bits = codeword_bits[
            :data_end_position
        ]

        return original_data_bits
